# Hernandez, Eujean Marie T. <br> CPE32S8

# Activity 3.1 - Recurrent Neural Networks

#### Objective(s):

This activity aims to introduce how to build a recurrent neural network

#### Intended Learning Outcomes (ILOs):
* Demonstrate how to build and train neural recurrent neural network
* Evaluate the score and accuracy of the recurrent neural network

#### Resources:
* Jupyter Notebook
* IMDB

#### Procedures
Load the necessary libraries

In [3]:
from __future__ import print_function
import keras
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Embedding
from keras.layers import SimpleRNN
from keras.datasets import imdb
from keras import initializers

* Set the max_features to 20000
* Set the maximum length of a sequence
* Use the batch size of 32

In [2]:
max_features = 20000
maxlen = 30
batch_size = 32

Load the data

In [29]:

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

25000 train sequences
25000 test sequences


 Truncates the sequences so that they are of the maximum length

In [30]:

x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

x_train shape: (25000, 30)
x_test shape: (25000, 30)


Check the example of sequence

In [31]:
x_train[123,:]

array([  219,   141,    35,   221,   956,    54,    13,    16,    11,
        2714,    61,   322,   423,    12,    38,    76,    59,  1803,
          72,     8, 10508,    23,     5,   967,    12,    38,    85,
          62,   358,    99], dtype=int32)

Build a recurrent neural network

In [32]:
rnn_hidden_dim = 5
word_embedding_dim = 50
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [33]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [34]:
import time

start_time = time.time()

model_rnn.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=10,
              validation_data=(x_test, y_test))

end_time = time.time()
runtime = end_time - start_time
print("Training runtime:", runtime, "seconds")


Epoch 1/10
782/782 [==============================] - 38s 47ms/step - loss: 0.5656 - accuracy: 0.7024 - val_loss: 0.4668 - val_accuracy: 0.7752
Epoch 2/10
782/782 [==============================] - 30s 38ms/step - loss: 0.4201 - accuracy: 0.8061 - val_loss: 0.4443 - val_accuracy: 0.7861
Epoch 3/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3810 - accuracy: 0.8268 - val_loss: 0.4560 - val_accuracy: 0.7881
Epoch 4/10
782/782 [==============================] - 29s 37ms/step - loss: 0.3584 - accuracy: 0.8417 - val_loss: 0.4577 - val_accuracy: 0.7874
Epoch 5/10
782/782 [==============================] - 27s 34ms/step - loss: 0.3417 - accuracy: 0.8525 - val_loss: 0.4486 - val_accuracy: 0.7884
Epoch 6/10
782/782 [==============================] - 27s 34ms/step - loss: 0.3270 - accuracy: 0.8602 - val_loss: 0.4621 - val_accuracy: 0.7857
Epoch 7/10
782/782 [==============================] - 27s 35ms/step - loss: 0.3128 - accuracy: 0.8684 - val_loss: 0.4937 - val_accuracy:

Evaluate the model using the test set.

In [35]:
score, acc = model_rnn.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 4s 5ms/step - loss: 0.5074 - accuracy: 0.7834
Test score: 0.5074489712715149
Test accuracy: 0.7833999991416931


Interpret the result

# The model achieve a low accuracy of 78.34% which indicates that the model is not learning meaningful patterns from the data.

#### Supplementary Activity

- Prepare the data to use sequences of length 80 rather than length 30.  Did it improve the performance?
- Try different values of the "max_features".  Can you improve the performance?
- Try smaller and larger sizes of the RNN hidden dimension.  How does it affect the model performance?  How does it affect the run time?

In [4]:
# type your code here


max_features = 30000
maxlen = 80
batch_size = 32

maxlen = 80

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

17464789/17464789 [==============================] - 2s 0us/step
25000 train sequences
25000 test sequences


In [5]:
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)


x_train shape: (25000, 80)
x_test shape: (25000, 80)


In [6]:
rnn_hidden_dim = 10
word_embedding_dim = 50
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [7]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [8]:
import time

start_time = time.time()

model_rnn.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=10,
              validation_data=(x_test, y_test))

end_time = time.time()
runtime = end_time - start_time
print("Training runtime:", runtime, "seconds")


Epoch 1/10
782/782 [==============================] - 90s 109ms/step - loss: 0.4935 - accuracy: 0.7409 - val_loss: 0.4269 - val_accuracy: 0.8290
Epoch 2/10
782/782 [==============================] - 84s 107ms/step - loss: 0.3196 - accuracy: 0.8639 - val_loss: 0.3484 - val_accuracy: 0.8456
Epoch 3/10
782/782 [==============================] - 71s 90ms/step - loss: 0.2711 - accuracy: 0.8884 - val_loss: 0.3517 - val_accuracy: 0.8448
Epoch 4/10
782/782 [==============================] - 77s 99ms/step - loss: 0.2379 - accuracy: 0.9038 - val_loss: 0.3933 - val_accuracy: 0.8402
Epoch 5/10
782/782 [==============================] - 73s 94ms/step - loss: 0.2138 - accuracy: 0.9171 - val_loss: 0.3928 - val_accuracy: 0.8414
Epoch 6/10
782/782 [==============================] - 76s 97ms/step - loss: 0.1914 - accuracy: 0.9279 - val_loss: 0.4570 - val_accuracy: 0.8324
Epoch 7/10
782/782 [==============================] - 74s 94ms/step - loss: 0.1716 - accuracy: 0.9366 - val_loss: 0.4425 - val_accurac

In [9]:
score, acc = model_rnn.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 8s 10ms/step - loss: 0.4593 - accuracy: 0.8328
Test score: 0.4592815637588501
Test accuracy: 0.8328400254249573


## Trying rnn_hidden_dim = 3

In [7]:
max_features = 30000
maxlen = 80
batch_size = 32

maxlen = 80

(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

25000 train sequences
25000 test sequences


In [8]:
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)


x_train shape: (25000, 80)
x_test shape: (25000, 80)


In [9]:
rnn_hidden_dim = 3
word_embedding_dim = 50
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, word_embedding_dim))  #This layer takes each integer in the sequence and embeds it in a 50-dimensional vector
model_rnn.add(SimpleRNN(rnn_hidden_dim,
                    kernel_initializer=initializers.RandomNormal(stddev=0.001),
                    recurrent_initializer=initializers.Identity(gain=1.0),
                    activation='relu',
                    input_shape=x_train.shape[1:]))

model_rnn.add(Dense(1, activation='sigmoid'))

In [10]:
rmsprop = keras.optimizers.RMSprop(lr = .0001)

model_rnn.compile(loss='binary_crossentropy',
              optimizer=rmsprop,
              metrics=['accuracy'])

In [13]:
import time

start_time = time.time()

model_rnn.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=10,
              validation_data=(x_test, y_test))

end_time = time.time()
runtime = end_time - start_time
print("Training runtime:", runtime, "seconds")


Epoch 1/10
782/782 [==============================] - 65s 84ms/step - loss: 0.2680 - accuracy: 0.8892 - val_loss: 0.4036 - val_accuracy: 0.8233
Epoch 2/10
782/782 [==============================] - 69s 89ms/step - loss: 0.2405 - accuracy: 0.9036 - val_loss: 0.3685 - val_accuracy: 0.8427
Epoch 3/10
782/782 [==============================] - 68s 87ms/step - loss: 0.2170 - accuracy: 0.9152 - val_loss: 0.3978 - val_accuracy: 0.8385
Epoch 4/10
782/782 [==============================] - 67s 86ms/step - loss: 0.1979 - accuracy: 0.9244 - val_loss: 0.4152 - val_accuracy: 0.8324
Epoch 5/10
782/782 [==============================] - 69s 88ms/step - loss: 0.1838 - accuracy: 0.9304 - val_loss: 0.4075 - val_accuracy: 0.8316
Epoch 6/10
782/782 [==============================] - 64s 82ms/step - loss: 0.1707 - accuracy: 0.9366 - val_loss: 0.4783 - val_accuracy: 0.8204
Epoch 7/10
782/782 [==============================] - 67s 86ms/step - loss: 0.1581 - accuracy: 0.9421 - val_loss: 0.5405 - val_accuracy:

In [14]:
score, acc = model_rnn.evaluate(x_test, y_test,
                            batch_size=batch_size)
print('Test score:', score)
print('Test accuracy:', acc)

782/782 [==============================] - 9s 11ms/step - loss: 0.5453 - accuracy: 0.8252
Test score: 0.5453377962112427
Test accuracy: 0.825160026550293


#type your answer here

**Prepare the data to use sequences of length 80 rather than length 30.  Did it improve the performance?** <br>
- Yes, increasing the sequence length from 30 to 80 improved the model's performance, with accuracy rising from 78.34% to 83.28%.


**Try different values of the "max_features".  Can you improve the performance?** <br>
- Increasing max_features to 30000  improved the model's performance because it considered a larger vocabulary, capturing more diverse patterns in the data.



**Try smaller and larger sizes of the RNN hidden dimension.  How does it affect the model performance?  How does it affect the run time?** <br>
- Increasing the size of the RNN hidden dimension increased the training time because the model needs more computation to learn and capture the patterns of the data.

#### Conclusion

## In this activity, I explored the construction and training of a recurrent neural network using the IMDB dataset. I learned that adjusting parameters such as sequence length, max_features, and hidden dimension layers can significantly impact the model's performance. After experimentation, I achieved an accuracy of 83.28% using a sequence length of 80, 30000 max features, and a hidden dimension of 10.

Google Colab Link: https://colab.research.google.com/drive/1fUz7hNGuuO3Yq0Y2hlBu4v8SE4710KNT?usp=sharing